src/TorchSharp/Tensor/Tensor.cs

```diff
@@ -881,6 +883,229 @@ public long[] size()
+            /// 
+            /// Refining is a special case of renaming that “lifts” unnamed dimensions.A None dim can be refined to have any name; a named dim can only be refined to have the same name.
+            /// Because named tensors can coexist with unnamed tensors, refining names gives a nice way to write named-tensor-aware code that works with both named and unnamed tensors.
+            /// names may contain up to one ellipsis argument, passed as "...". The ellipsis is expanded greedily; it is expanded in-place to fill names to the same length as
+            /// this.shape using names from the corresponding indices of this.names.
+            /// </summary>
+            /// <param name="names">A list of names, one for each dimension in the tensor.</param>
+            /// <remarks>The named tensor API is experimental and subject to change.</remarks>
+            public Tensor refine_names(IEnumerable<string> names)
+            {
+                var dimNamesArray = ExpandEllipsis(names);
+
+                using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                IntPtr namesRef = pinnedArray.CreateArray(dimNamesArray);
+
+                IntPtr res = THSTensor_refine_names(Handle, namesRef, dimNamesArray.Length);
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                return new Tensor(res);
+            }
+
+            private IntPtr[] ExpandEllipsis(IEnumerable<string?> names)
+            {
+                var namesArray = names.ToArray();
+                IntPtr[] dimNamesArray = new IntPtr[ndim];
+
+                if (namesArray.Contains("...")) {
+                    int idx = -1;
+                    for (int i = 0; i < namesArray.Length; i++)
+                        if (namesArray[i] == "...") { idx = i; break; }
+
+                    if (idx == namesArray.Length - 1) { // '...' appears last.
+
+                        int i = 0;
+                        for (; i < idx; i++)
+                            dimNamesArray[i] = MarshalDimensionString(namesArray[i]);
+
+                        if (has_names()) {
+
+                            var n = this.names.ToArray();
+
+                            for (; i < dimNamesArray.Length; i++)
+                                dimNamesArray[i] = MarshalDimensionString(n[i]);
+
+                        } else {
+                            for (; i < dimNamesArray.Length; i++)
+                                dimNamesArray[i] = IntPtr.Zero;
+                        }
+
+                    } else { // Names before and after.
+
+                        int dIdx = 0;
+                        for (int i = 0; i < idx; i++, dIdx++)
+                            dimNamesArray[dIdx] = MarshalDimensionString(namesArray[i]);
+
+                        int missing_dims = (int)(ndim - namesArray.Length + 1);
+
+                        if (has_names()) {
+
+                            var n = this.names.ToArray();
```

Comment:
> var n = this.names.ToArray(); [](http://example.com/codeflow?start=27&length=30)

her too we can git id of this allocation too I guess.

Reply:
Yup.